### PR TITLE
[Network] App Gateway WAF Ignite features about WAF policy

### PR DESCRIPTION
**Pull request by haroldrandom**
* Add feature requested https://github.com/Azure/azure-cli/issues/11046 
  - Add new command subgroup **managed-rules** to waf-policy
  - Add new command subgroup **policy-setting** to waf-policy
  - [BREAKING CHANGE] **waf-policy rule** rename to **waf-policy custom-rule**
  - Add --firewall-policy to application-gateway url-path-map rule create
  - Add --firewall-policy to application-gateway http-listener
* Fix https://github.com/Azure/azure-cli/issues/10897
---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

File: src/azure-cli/azure/cli/command_modules/network/custom.py [hunk 1]
```diff
@@ -341,14 +341,15 @@ def create_ag_http_listener(cmd, resource_group_name, application_gateway_name,
         host_name=host_name,
         require_server_name_indication=True if ssl_cert and host_name else None,
         protocol='https' if ssl_cert else 'http',
-        ssl_certificate=SubResource(id=ssl_cert) if ssl_cert else None)
+        ssl_certificate=SubResource(id=ssl_cert) if ssl_cert else None,
+        firewall_policy=SubResource(id=firewall_policy) if firewall_policy else None)
```
Comment (MyronFanQiu) [inline src/azure-cli/azure/cli/command_modules/network/custom.py]: if cmd.supported_api_version(min_api='2019-09-01'):

File: src/azure-cli/azure/cli/command_modules/network/custom.py [hunk 2]
```diff
@@ -912,7 +918,8 @@ def create_ag_url_path_map_rule(cmd, resource_group_name, application_gateway_na
         name=item_name,
         paths=paths,
         backend_address_pool=SubResource(id=address_pool) if address_pool else default_backend_pool,
-        backend_http_settings=SubResource(id=http_settings) if http_settings else default_http_settings)
+        backend_http_settings=SubResource(id=http_settings) if http_settings else default_http_settings,
+        firewall_policy=SubResource(id=firewall_policy) if firewall_policy else None)
```
Comment (MyronFanQiu) [inline src/azure-cli/azure/cli/command_modules/network/custom.py]: if cmd.supported_api_version(min_api='2019-09-01'):?

File: src/azure-cli/azure/cli/command_modules/network/commands.py [hunk 3]
```diff
@@ -455,29 +455,63 @@ def _make_singular(value):
...
-        g.custom_command('create', 'create_ag_waf_rule')
-        g.custom_command('delete', 'delete_ag_waf_rule')
-        g.custom_command('list', 'list_ag_waf_rules')
-        g.custom_show_command('show', 'show_ag_waf_rule')
-        g.generic_update_command('update', command_type=network_ag_waf_sdk,
+    with self.command_group('network application-gateway waf-policy policy-setting', network_ag_waf_sdk,
+                            client_factory=cf_app_gateway_waf_policy,
+                            min_api='2019-09-01') as g:
+        g.custom_command('list', 'list_waf_policy_setting')
+        g.generic_update_command('update',
```
Comment (MyronFanQiu) [inline src/azure-cli/azure/cli/command_modules/network/commands.py]: Is there a create command?

File: src/azure-cli/azure/cli/command_modules/network/custom.py [hunk 4]
```diff
@@ -1103,37 +1141,156 @@ def delete_ag_waf_rule(cmd, client, resource_group_name, policy_name, rule_name,
...
+    for rule_set in waf_policy.managed_rules.managed_rule_sets:
+        if rule_set.rule_set_type == rule_set_type:
+            for rule_override in rule_set.rule_group_overrides:
+                if rule_override.rule_group_name == rule_group_name:
+                    rule_override.rules.extend(managed_rule_overrides)
+                    break
+            else:
+                rule_set.rule_group_overrides.append(rule_group_override)
+            break
+    else:
```
Comment (MyronFanQiu) [inline src/azure-cli/azure/cli/command_modules/network/custom.py]: Although the logic is correct, it's confusing semantically...

File: src/azure-cli/azure/cli/command_modules/network/custom.py [hunk 5]
```diff
@@ -1103,37 +1141,156 @@ def delete_ag_waf_rule(cmd, client, resource_group_name, policy_name, rule_name,
...
+    for rule_set in instance.managed_rules.managed_rule_sets:
+        if rule_set.rule_set_type == rule_set_type:
+            for rule_group in rule_set.rule_group_overrides:
+                if rule_group.rule_group_name == rule_group_name:
+                    rule_group.rules = managed_rule_overrides
+                    break
+            else:
+                rule_set.rule_group_overrides.append(rule_group_override)
+            break
+    else:
```
Comment (MyronFanQiu) [inline src/azure-cli/azure/cli/command_modules/network/custom.py]: I don't think update should be able to add new rule set

File: src/azure-cli/azure/cli/command_modules/network/tests/latest/test_network_commands.py [hunk 6]
```diff
@@ -946,52 +946,314 @@ def test_network_app_gateway_waf_v2_config(self, resource_group):
...
+                 '--priority 100 --action log --rule-type MatchRule')
+
+        # update some policy settings of this waf-policy
+        self.cmd('network application-gateway waf-policy policy-setting update -g {rg} --policy-name {waf} '
+                 '--state Enabled --file-upload-limit-in-mb 64 --mode Prevention')
+
+        # add two managed rule set to the managed rules of this waf-policy
+        self.cmd('network application-gateway waf-policy managed-rule rule-set add -g {rg} --policy-name {waf} '
+                 '--type OWASP --version 3.0 '
+                 '--group-name {csr_grp1} --rules 921100 921110')
```
Comment (MyronFanQiu) [inline src/azure-cli/azure/cli/command_modules/network/tests/latest/test_network_commands.py]: How can user know 92110 921110?

Comment (haroldrandom) [reply src/azure-cli/azure/cli/command_modules/network/tests/latest/test_network_commands.py]: No. user can not. May be we should provide another like `list-managed-rule-set` to show all of this thing?

File: src/azure-cli/azure/cli/command_modules/network/custom.py (hunk 2 above)
Comment (haroldrandom) [reply src/azure-cli/azure/cli/command_modules/network/custom.py]: You are right. Fixing it.

File: src/azure-cli/azure/cli/command_modules/network/custom.py (hunk 1 above)
Comment (haroldrandom) [reply src/azure-cli/azure/cli/command_modules/network/custom.py]: You are right. Fixing it.

File: src/azure-cli/azure/cli/command_modules/network/tests/latest/test_network_commands.py (hunk 6 above)
Comment (MyronFanQiu) [reply src/azure-cli/azure/cli/command_modules/network/tests/latest/test_network_commands.py]: list-managed-rule? How do you know these values? It's better to provide a command. At least, we can provide a link.

Comment (haroldrandom) [reply src/azure-cli/azure/cli/command_modules/network/tests/latest/test_network_commands.py]: `list-managed-rule-set` There is no such a link to list all managed-rule-set. I gave a link https://docs.microsoft.com/en-us/azure/web-application-firewall/afds/afds-overview But it doesn't help a lot I think.

There is a command `az network application-gateway waf-config list-rule-sets` doing something similar.

For now. user have to go to portal to find out which rule-set and rules he/she want's to disable.
By default, all of rules in a rule-set are Enabled.

File: src/azure-cli/azure/cli/command_modules/network/_params.py [hunk 7]
```diff
@@ -371,38 +373,101 @@ def load_arguments(self, _):
...
         c.argument('policy_name', options_list='--policy-name', id_part=None)
 
-    with self.argument_context('network application-gateway waf-policy rule match-condition', min_api='2018-12-01') as c:
+    with self.argument_context('network application-gateway waf-policy custom-rule match-condition',
+                               min_api='2018-12-01') as c:
         c.argument('operator', arg_type=get_enum_type(WebApplicationFirewallOperator), help='Operator for matching.')
-        c.argument('negation_condition', options_list='--negate', arg_type=get_three_state_flag(), help='Match the negative of the condition.')
-        c.argument('match_values', options_list='--values', nargs='+', help='Space-separated list of values to match.')
-        c.argument('transforms', arg_type=get_enum_type(WebApplicationFirewallTransform), nargs='+', help='Space-separated list of transforms to apply when matching.')
+        c.argument('negation_condition',
```
[the remark this answers is not in the document]
Comment (haroldrandom) [reply src/azure-cli/azure/cli/command_modules/network/_params.py]: I just fold it. change nothing here.

File: src/azure-cli/azure/cli/command_modules/network/_params.py [hunk 8]
```diff
@@ -218,6 +218,7 @@ def load_arguments(self, _):
         c.argument('ssl_cert', help='The name or ID of the SSL certificate to use.', completer=get_ag_subresource_completion_list('ssl_certificates'))
         c.ignore('protocol')
         c.argument('host_name', help='Host name to use for multisite gateways.')
+        c.argument('firewall_policy', min_api='2019-09-01', help='Reference to the FirewallPolicy resource.')
```
[the remark this answers is not in the document]
Comment (haroldrandom) [reply src/azure-cli/azure/cli/command_modules/network/_params.py]: Fixed

File: src/azure-cli/azure/cli/command_modules/network/commands.py (hunk 3 above)
Comment (haroldrandom) [reply src/azure-cli/azure/cli/command_modules/network/commands.py]: Can not create. Update only.

File: src/azure-cli/azure/cli/command_modules/network/custom.py (hunk 5 above)
Comment (haroldrandom) [reply src/azure-cli/azure/cli/command_modules/network/custom.py]: Fix it.

Comment (haroldrandom) [reply src/azure-cli/azure/cli/command_modules/network/custom.py]: Raise CLIError instead.

File: src/azure-cli/azure/cli/command_modules/network/tests/latest/test_network_commands.py (hunk 6 above)
Comment (haroldrandom) [reply src/azure-cli/azure/cli/command_modules/network/tests/latest/test_network_commands.py]: I found a link https://docs.microsoft.com/en-us/azure/web-application-firewall/ag/application-gateway-crs-rulegroups-rules for user to query rule sets and added it to waf-policy managed-rule rule-set add/update.

We should better need a list command to show all these rules in the feature.

File: src/azure-cli/azure/cli/command_modules/network/_params.py [hunk 9]
```diff
@@ -292,7 +292,7 @@ def load_arguments(self, _):
         c.argument('paths', nargs='+', help='Space-separated list of paths to associate with the rule. Valid paths start and end with "/" (ex: "/bar/")', arg_group='First Rule')
         c.argument('address_pool', help='The name or ID of the backend address pool to use with the created rule.', completer=get_ag_subresource_completion_list('backend_address_pools'), arg_group='First Rule')
         c.argument('http_settings', help='The name or ID of the HTTP settings to use with the created rule.', completer=get_ag_subresource_completion_list('backend_http_settings_collection'), arg_group='First Rule')
-        c.argument('firewall_policy', min_api='2019-09-01', help='Reference to the FirewallPolicy resource.')
+        c.argument('firewall_policy', min_api='2019-09-01', help='Name or ID of a Firewall Policy resource.')
```
Comment (MyronFanQiu) [inline src/azure-cli/azure/cli/command_modules/network/_params.py]: If we support Name of the firewall policy, we should build valid resource ID for user.

File: src/azure-cli/azure/cli/command_modules/network/custom.py [hunk 10]
```diff
@@ -341,8 +341,11 @@ def create_ag_http_listener(cmd, resource_group_name, application_gateway_name,
         host_name=host_name,
         require_server_name_indication=True if ssl_cert and host_name else None,
         protocol='https' if ssl_cert else 'http',
-        ssl_certificate=SubResource(id=ssl_cert) if ssl_cert else None,
-        firewall_policy=SubResource(id=firewall_policy) if firewall_policy else None)
+        ssl_certificate=SubResource(id=ssl_cert) if ssl_cert else None)
+
+    if cmd.supported_api_version(min_api='2019-09-01'):
+        new_listener.firewall_policy = SubResource(id=firewall_policy) if firewall_policy else None
```
Comment (MyronFanQiu) [inline src/azure-cli/azure/cli/command_modules/network/custom.py]: here. firewall_policy must be a resource ID.

File: src/azure-cli/HISTORY.rst [hunk 11]
```diff
@@ -29,6 +29,14 @@ Release History
 
 * Fix #11121: When using `az keyvault certificate list`, passing `--include-pending` now doesn't require a value of `true` or `false`
 
+**Network**
+
+* application-gateway waf-policy: Add subgroup managed-rules to manage managed rule sets and exclusion rules
+* application-gateway waf-policy: Add subgroup policy-setting to manage global configuration of a waf-policy
+* [BREAKING CHANGE] application-gateway waf-policy: Rename subgroup rule to custom-rule
```
[the remark this answers is not in the document]
Comment (haroldrandom) [reply src/azure-cli/HISTORY.rst]: As talked offline.

File: src/azure-cli/azure/cli/command_modules/network/custom.py (hunk 10 above)
Comment (haroldrandom) [reply src/azure-cli/azure/cli/command_modules/network/custom.py]: Fix it in _validator.py

File: src/azure-cli/azure/cli/command_modules/network/_params.py (hunk 9 above)
Comment (haroldrandom) [reply src/azure-cli/azure/cli/command_modules/network/_params.py]: Fix it in _validator.py

**Pull request merged**